synapse/lib/lmdbslab.py

```diff
@@ -1111,3 +1168,87 @@ def bump(self):
         if not self.bumped:
             self.curs.close()
             self.bumped = True
+
+class ScanBack(Scan):
+    '''
+    A state-object used by Slab.  Not to be instantiated directly.
+
+    Scans backwards.
+    '''
+    def first(self):
+
+        if not self.curs.last():
+            return False
+
+        if self.dupsort:
+            self.iterfunc = functools.partial(lmdb.Cursor.iterprev_dup, keys=True)
+        else:
+            self.iterfunc = lmdb.Cursor.iterprev
+
+        self.genr = self.curs.iterprev()
+
+        self.atitem = next(self.genr)
+        return True
+
+    def set_key(self, lkey):
+
+        if not self.curs.set_key(lkey):
+            return False
+
+        # set_key for a scan is only logical if it's a dup scan
+        if self.dupsort:
+            if not self.curs.last_dup():
+                return False
+
+            self.iterfunc = functools.partial(lmdb.Cursor.iterprev_dup, keys=True)
+        else:
+            self.iterfunc = lmdb.Cursor.iterprev
+
+        self.genr = self.iterfunc(self.curs)
+        self.atitem = next(self.genr)
+        return True
+
+    def set_range(self, lkey):
+
+        if not self.curs.set_range(lkey):
+            if not self.curs.last():
+                return False
+
+        self.iterfunc = lmdb.Cursor.iterprev
+
+        self.genr = self.iterfunc(self.curs)
+        self.atitem = next(self.genr)
+
+        if not self.atitem[0] <= lkey:
+            self.atitem = next(self.genr)
+
+        return True
+
+    def iternext(self):
```

Comment:
why is this any different than parent method?

Reply:
It is a little bit different in how it handles being bumped. After calling set_range, if self.curs.item() != self.atitem we will still want to call next(self.genr) since we would have an item that is further back in the scan, rather than already being at the next item in the scan when going forwards.